src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1525,11 +1519,11 @@ protected internal override void OnProjectRemoved(ProjectId projectId)
 
                 base.OnProjectRemoved(projectId);
 
-                _deferredJoinableTasks.Add(_threadingContext.JoinableTaskFactory.RunAsync(async () =>
+                _threadingContext.RunWithShutdownBlockAsync(async cancellationToken =>
```

Comment:
1. i like that this is explicit.
2. i dont' particularly like that we're explicitly stating: this will block shutdown.  I'm struggling to think of any scenarios where we ever want to block shutdown.

Reply:
Because VS shuts the CLR down before terminating the process, it's very important that no managed code is executing during that time or a process crash is likely. So we block shutdown till (hopefully) all managed code reports completion.

Reply:
> Because VS shuts the CLR down before terminating the process, it's very important that no managed code is executing during that time or a process crash is likely.

But the process is being torn down.  So i don't care if anything crashes.  It seems like we're just adding a cost, to make something expensive taht users would always want to effectively be instantaneous.

In other words:  i close VS by killing it in task-manager.  Doing work to make us slower than that seems wasteful.

Reply:
When VS crashes on shutdown, the users can tell. And Windows files a Watson report which slows them down.
That said, I learned today that VS no longer shuts down the CLR on customer machines. It only does it on our optprof machines when we're running certain test runs, because without shutting things down properly, we can't get the optimization data we need. Crashes while shutting down then blocks those runs, so we want to keep them happy.

Reply:
> That said, I learned today that VS no longer shuts down the CLR on customer machines. It only does it on our optprof machines when we're running certain test runs, because without shutting things down properly, we can't get the optimization data we need. Crashes while shutting down then blocks those runs, so we want to keep them happy.

Heh, that's news to me too. @AArnott Should we only be blocking if this is an OptProf run then? Why slow down customer paths for internal tooling?

Reply:
Good question, @jasonmalinowski. And unfortunately, results in a rather complicated non-answer, based on my conversation last week with the P&R team.
At the platform level we *do* want to support blocking shutdown on completion of async tasks so that the process doesn't terminate while persistent storage or network writes are ongoing to avoid corrupting data. At the feature level, this news _almost_ gives you the freedom to decide to only track those tasks that you think *should* block shutdown, and leave the rest of them untracked so they don't need to complete or cancel before shutdown. 
The chief concern with this is optprof is crucial, and if it exercises code paths that don't run at any other time it's more likely to fail. In particular if there are deadlocks on shutdown because you're blocking on more code during an optprof run than otherwise, those may be relatively expensive bugs to investigate and fix, particularly if mitigating those deadlocks requires deep changes to your work scheduling policies. So it is useful to do it all the time to encourage healthier programming patterns in general and to keep optprof runs healthy.
I am not settled that the current policy is the best one, and I hope we can strike a better deal in some way, but we haven't found it yet.